### ISSUE TITLE
Creating TomlObject Trees

### ISSUE THREAD

**Issue by pragmatrix** [paiden/Nett#21]
Hi, I'd like to use Nett to create toml files based on a custom configuration representation, but I can't find anything in the public API that helps me to create a proper TomlObject based tree.

I see some helpers in `TomlTableExtensions` for adding values to a table, but they don't seem to support the addition of booleans, or an array of strings for example.

Do I miss something?

**Comment by paiden** [comment]
No you didn't miss anything (as far as I can tell after a quick investigation). 

Simply put, I forgot to implement the remaining overloads of these helpers, because for my internal projects using the Lib, I only use strongly typed CLR objects for the read/write methods.

e.g. 

```
public class Config { public bool X {get; set; } }

var cfg = new Config();
Toml.WriteFile(cfg, "test.toml");
cfg = Toml.ReadFile<Config>("test.toml");
```

I will add the missing methods within the next few days.

**Comment by pragmatrix** [comment]
May be it would be best to retire these Add methods (except one that takes adds a TomlObject perhaps) and instead add factory methods to the TomlObject abstract class that create TomlObject instances from intrinsic types. TomlObject because - as I can see - the creation of TomlObjects is not possible without a reference to the internal ITomlRoot of an existing TomlObject.

For collection types like TomlArray and TomlTable, we could provide one factory method each that takes in TomlObject sequences or a sequence of `Tuple<string, TomlObject>` for example.

**Comment by paiden** [comment]
I'm not sure I understand why you would like to retire the add methods? What is the benefit you see in those factory methods?

From a client usage API perspective I think 
```
tbl.Add("x", 100);
```
would be cleaner and more convenient than 
```
tbl.Add("x", tbl.Create(100));
```

Currently error cases like this
```
var tbl2 = ...
tbl.Add("x", tbl2.Create(100); 
```
are not possible with the current API design (as Add(string, TomlObject) is internal; only the this[string] setter is a special case where you can set an object with an invalid root).

**Comment by pragmatrix** [comment]
Yes sure, you are right. The Add methods can stay of course for a convenient way to add rows to a table. I guess I overstepped here because they were incomplete and not useful for my intensions.

Having the factory methods is required to separate construction from adding. This way a complete tree can be built by a simple recursive function ([like here](https://gist.github.com/pragmatrix/498a0d01db194f70978352beac64e9be#file-configuration-toml-fs-L32)), and is not dependent on which container the items need to be added or inserted (for example, a separate set of Add methods are then needed for arrays).

And regarding the error case, the factory API should check of root inconsistencies. I can try to cover that.

Also I find the naming `Create` somehow an unlucky choice. May be `TomlObject.CreateFor(intrinsic value)` would sound a bit better.

**Comment by paiden** [comment]
While fixing issue #23 I found out why these factory methods you describe are indeed useful. 

BTW: Thanks a lot for giving my lib a try and reporting all these issues. My internal use cases seem to be a lot simpler than yours :)

E.g at the moment the indexer of the TOML table is completely useless, as it is (nearly)impossible to insert/set something with it, as no 'unrooted' object can be created (the setter already checks if the root is correct). Your factory methods will enable to do so:

```
tbl["x"] = tbl.Create(100);
```
My current plan is to keep the add methods including the missing overloads and add the factory methods also.

**Comment by paiden** [comment]
I've added the new factory methods.

I took your PR #22 as a base and incorporated some Nett principles and tried to come up with a nice to use API. 

For signaling the user, an created object belongs to the graph it was created on I choose to use the suffix 'Attached'. There are a lot of overloads to ensure the user doesn't have do a lot of casting just because he uses standard .Net types. Especially in the IEnumerable<...> cases.

Also I separated the methods into various set of Create/Attach method sets. I chose these sets so that overload resolution problems should not happen accidentally when using the API to construct object graphs (things that happened to me while writing unit tests while the sets were not there yet).

I would like to know what you think about the new API. For now I'm quite happy with it overall and think it could get integrated into 'master' the way it is.
